pkg/rpcbackend/backend.go

```diff
@@ -160,63 +313,89 @@ func (rc *RPCClient) SyncRequest(ctx context.Context, rpcReq *RPCRequest) (rpcRe
 		}()
 	}
 
-	// We always set the back-end request ID - as we need to support requests coming in from
-	// multiple concurrent clients on our front-end that might use clashing IDs.
-	var beReq = *rpcReq
-	beReq.JSONRpc = "2.0"
-	rpcTraceID := rc.allocateRequestID(&beReq)
-	if rpcReq.ID != nil {
-		// We're proxying a request with front-end RPC ID - log that as well
-		rpcTraceID = fmt.Sprintf("%s->%s", rpcReq.ID, rpcTraceID)
-	}
+	if rc.requestBatchQueue != nil {
+		req := &batchRequest{
+			rpcReq: rpcReq,
+			rpcRes: make(chan *RPCResponse, 1),
+			rpcErr: make(chan error, 1),
+		}
 
-	rpcRes = new(RPCResponse)
+		select {
+		case rc.requestBatchQueue <- req:
+		case <-ctx.Done():
+			err := i18n.NewError(ctx, signermsgs.MsgRequestCanceledContext, rpcReq.ID)
+			return RPCErrorResponse(err, rpcReq.ID, RPCCodeInternalError), err
+		}
 
-	log.L(ctx).Debugf("RPC[%s] --> %s", rpcTraceID, rpcReq.Method)
-	if logrus.IsLevelEnabled(logrus.TraceLevel) {
-		jsonInput, _ := json.Marshal(rpcReq)
-		log.L(ctx).Tracef("RPC[%s] INPUT: %s", rpcTraceID, jsonInput)
-	}
-	rpcStartTime := time.Now()
-	res, err := rc.client.R().
-		SetContext(ctx).
-		SetBody(beReq).
-		SetResult(&rpcRes).
-		SetError(rpcRes).
-		Post("")
-
-	// Restore the original ID
-	rpcRes.ID = rpcReq.ID
-	if err != nil {
-		err := i18n.NewError(ctx, signermsgs.MsgRPCRequestFailed, err)
-		log.L(ctx).Errorf("RPC[%s] <-- ERROR: %s", rpcTraceID, err)
-		rpcRes = RPCErrorResponse(err, rpcReq.ID, RPCCodeInternalError)
-		return rpcRes, err
-	}
-	if logrus.IsLevelEnabled(logrus.TraceLevel) {
-		jsonOutput, _ := json.Marshal(rpcRes)
-		log.L(ctx).Tracef("RPC[%s] OUTPUT: %s", rpcTraceID, jsonOutput)
-	}
-	// JSON/RPC allows errors to be returned with a 200 status code, as well as other status codes
-	if res.IsError() || rpcRes.Error != nil && rpcRes.Error.Code != 0 {
-		rpcMsg := rpcRes.Message()
-		errLog := rpcMsg
-		if rpcMsg == "" {
-			// Log the raw result in the case of JSON parse error etc. (note that Resty no longer
-			// returns this as an error - rather the body comes back raw)
-			errLog = string(res.Body())
-			rpcMsg = i18n.NewError(ctx, signermsgs.MsgRPCRequestFailed, res.Status()).Error()
-		}
-		log.L(ctx).Errorf("RPC[%s] <-- [%d]: %s", rpcTraceID, res.StatusCode(), errLog)
-		err := fmt.Errorf(rpcMsg)
-		return rpcRes, err
-	}
-	log.L(ctx).Infof("RPC[%s] <-- %s [%d] OK (%.2fms)", rpcTraceID, rpcReq.Method, res.StatusCode(), float64(time.Since(rpcStartTime))/float64(time.Millisecond))
-	if rpcRes.Result == nil {
-		// We don't want a result for errors, but a null success response needs to go in there
-		rpcRes.Result = fftypes.JSONAnyPtr(fftypes.NullString)
+		select {
+		case rpcRes := <-req.rpcRes:
+			return rpcRes, nil
+		case err := <-req.rpcErr:
+			return nil, err
+		case <-ctx.Done():
+			err := i18n.NewError(ctx, signermsgs.MsgRequestCanceledContext, rpcReq.ID)
+			return RPCErrorResponse(err, rpcReq.ID, RPCCodeInternalError), err
+		}
+	} else {
```

Comment:
Just a style thing, but maybe have a think about splitting into separate functions as we have a big if/else

Reply:
like:

```go
if rc.requestBatchQueue != nil {
    return rc.batchSyncRequest(ctx, rpcReq)
}
// rest of function goes here unchanged from before the PR
```

Reply:
great point. Addressed.